### PR TITLE
Create Features.Pooling

### PR DESCRIPTION
**Pull request by Thundermaker300**
Made for creating our own pools for classes. Has support for NorthwoodLib's initial pools (ListPool, HashsetPool, StringBuilderPool), as well as additions (currently only DictPool)

Benefits:
* Documentation: New users will understand what pools are and how to use them in their code, optimizing their plugins (even slightly, it all helps).
* Less references: Plugin developers are much less likely to use pools due to it requiring the NorthwoodLib reference. Even less new programmers will even take a glance at the NorthwoodLib dll to begin with. Why would they add a new reference to their plugin for such a little thing, even though it goes a long way? This fixes that.
* Separates our pools from Northwood's pools. If a developer uses pools incorrectly (retrieves lists but doesn't return them), it will not affect the micro-performance of the game itself.

Downsides:
* Since our pool is separate from Northwood, any objects stored by the game will not be used in plugins.

### CODE REVIEW

**Review by IRacle1 — COMMENTED** [review]
Personally i dont like the `DictPool` class name.
Also you can add the `Rent(int capacity)` and `Rent(Dictionary<T, V>)` method like in NW's list pool

**Comment by Thundermaker300** [comment]
> Personally i dont like the `DictPool` class name.
What would you like the name to be?

> Also you can add the `Rent(int capacity)` and `Rent(Dictionary<T, V>)` method like in NW's list pool
I looked, although I'm not sure how to set the capacity of a dictionary. I'll add Rent(Dictionary) though.

**Comment by Thundermaker300** [comment]
> first: you can set capacity using ctor
Yes I know, but how do you expand the capacity of currently existing dictionaries? I mean that's the whole point of a pool right

**Comment by IRacle1** [comment]
> > first: you can set capacity using ctor
> 
> Yes I know, but how do you expand the capacity of currently existing dictionaries? I mean that's the whole point of a pool right

iirc, you cant change dictionary capacity😢

**Comment by iopietro** [comment]
I do like `DictionaryPool` too.

File: Exiled.API/Features/Pools/StringBuilderPool.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        /// <summary>
+        /// Returns the contents of the <see cref="StringBuilder"/> and returns it to the pool.
+        /// </summary>
+        /// <param name="obj">The <see cref="StringBuilder"/> to return.</param>
+        /// <returns>The value of the <see cref="StringBuilder"/>.</returns>
+        public string ToStringReturn(StringBuilder obj)
+        {
+            string s = obj.ToString();
+            Return(obj);
+            return s;
```
Comment (iopietro) [inline Exiled.API/Features/Pools/StringBuilderPool.cs:66]:
```suggestion
            string s = obj.ToString();
            
            Return(obj);
            
            return s;
```

File: Exiled.API/Features/Roles/Scp939Role.cs [hunk 2]
```diff
@@ -65,6 +67,14 @@ internal Scp939Role(Scp939GameRole baseRole)
             MimicryRecorder = mimicryRecorder;
         }
 
+        /// <summary>
+        /// Finalizes an instance of the <see cref="Scp939Role"/> class.
+        /// </summary>
+        ~Scp939Role()
+        {
+            ListPool<Player>.Pool.Return(VisiblePlayers);
+        }
```
Comment (iopietro) [inline Exiled.API/Features/Roles/Scp939Role.cs:76]:
```suggestion
        ~Scp939Role() => ListPool<Player>.Pool.Return(VisiblePlayers);
```

File: Exiled.CustomRoles/API/Features/CustomRole.cs [hunk 3]
```diff
@@ -544,7 +543,7 @@ public bool TryAddFriendlyFire(RoleTypeId roleToAdd, float ffMult)
         /// <returns> Whether the item was able to be added. </returns>
         public bool TryAddFriendlyFire(Dictionary<RoleTypeId, float> ffRules, bool overwrite = false)
         {
-            Dictionary<RoleTypeId, float> temporaryFriendlyFireRules = new();
+            Dictionary<RoleTypeId, float> temporaryFriendlyFireRules = DictionaryPool<RoleTypeId, float>.Pool.Get();
             foreach (KeyValuePair<RoleTypeId, float> roleFF in ffRules)
```
Comment (iopietro) [inline Exiled.CustomRoles/API/Features/CustomRole.cs:547]:
```suggestion
            Dictionary<RoleTypeId, float> temporaryFriendlyFireRules = DictionaryPool<RoleTypeId, float>.Pool.Get();

            foreach (KeyValuePair<RoleTypeId, float> roleFF in ffRules)
```

File: Exiled.API/Features/Core/EObject.cs [hunk 4]
```diff
@@ -540,14 +540,17 @@ public static T FindActiveObjectOfType<T>(Func<EObject, bool> predicate)
...
+            List<T> objects = ListPool<T>.Pool.Get();
             foreach (EObject @object in InternalObjects.Where(predicate))
             {
                 if (@object.Cast(out T obj))
                     objects.Add(obj);
             }
 
-            return objects.ToArray();
+            T[] array = objects.ToArray();
+            ListPool<T>.Pool.Return(objects);
```
Comment (iopietro) [inline Exiled.API/Features/Core/EObject.cs:551]: Should we add a `ListPool<T>.Pool.ToArrayReturn(...`)?

**Review by iopietro — DISMISSED** [review]

File: Exiled.API/Features/Core/EObject.cs (hunk 4 above)
Comment (Thundermaker300) [reply Exiled.API/Features/Core/EObject.cs:551]: Good idea, will add later today

**Review by iopietro — APPROVED** [review]

**Review by IRacle1 — APPROVED** [review]

**Review by iopietro — APPROVED** [review]

**Pull request merged**